File: backend/app/collectors/dns_collector.py
import socket
import time


def resolve(host: str) -> dict:
    result = {"a_records": [], "resolution_ms": None, "error": None}
    try:
        t0 = time.perf_counter()
        addr_info = socket.getaddrinfo(host, None, socket.AF_UNSPEC, socket.SOCK_STREAM)
        result["resolution_ms"] = round((time.perf_counter() - t0) * 1000)
        result["a_records"] = sorted({info[4][0] for info in addr_info})
    except Exception as exc:
        result["error"] = str(exc)[:512]
    return result

File: backend/src/collectors/dnsCollector.ts
import * as dns from "dns";

export interface DnsResult {
  a_records: string[];
  resolution_ms: number | null;
  error: string | null;
}

export function resolve(host: string): Promise<DnsResult> {
  return new Promise((resolve) => {
    const result: DnsResult = { a_records: [], resolution_ms: null, error: null };
    const start = Date.now();

    dns.lookup(host, { all: true }, (err, addresses) => {
      if (err) {
        result.error = err.message;
        return resolve(result);
      }
      result.resolution_ms = Date.now() - start;
      result.a_records = [...new Set(addresses.map((a) => a.address))].sort();
      resolve(result);
    });
  });
}
